src/Tools/Source/CompilerGeneratorTools/Source/CSharpSyntaxGenerator/SourceWriter.cs

```diff
@@ -736,34 +711,51 @@ private void WriteGreenFactory(Node nd, bool withSyntaxFactoryContext = false)
 
                 if (!IsAnyList(field.Type) && !IsOptional(field))
                 {
-                    WriteLine("      if ({0} == null)", CamelCase(field.Name));
-                    WriteLine("        throw new ArgumentNullException(nameof({0}));", CamelCase(field.Name));
+                    WriteLine("if ({0} == null) throw new ArgumentNullException(nameof({0}));", CamelCase(field.Name));
                 }
                 if (field.Type == "SyntaxToken" && field.Kinds != null && field.Kinds.Count > 0)
                 {
                     if (IsOptional(field))
                     {
-                        WriteLine("      if ({0} != null)", CamelCase(field.Name));
-                        WriteLine("      {");
-                    }
-                    WriteLine("      switch ({0}.Kind)", pname);
-                    WriteLine("      {");
-                    foreach (var kind in field.Kinds)
-                    {
-                        WriteLine("        case SyntaxKind.{0}:", kind.Name);
+                        WriteLine("if ({0} != null)", CamelCase(field.Name));
+                        OpenBlock();
                     }
-                    //we need to check for Kind=None as well as node == null because that's what the red factory will pass
-                    if (IsOptional(field))
+
+                    if (field.Kinds.Count == 1 && !IsOptional(field))
                     {
-                        WriteLine("        case SyntaxKind.None:");
+                        WriteLine("if ({0}.Kind != SyntaxKind.{1}) throw new ArgumentException(nameof({0}));", pname, field.Kinds[0].Name);
```

Comment:
special cases switches where we only have a single case.  In that case (no pun intended) we only use an 'if' check for consistency with the null checks.